src/aggregators/extrande.jl

```diff
@@ -0,0 +1,140 @@
+# Define the aggregator. 
+struct Extrande <: AbstractAggregatorAlgorithm end
+
+"""
+Extrande sampling method for jumps with defined rate bounds.
+"""
+
+nullaffect!(integrator) = nothing
+const NullAffectJump = ConstantRateJump((u, p, t) -> 0.0, nullaffect!)
+
+mutable struct ExtrandeJumpAggregation{T, S, F1, F2, F3, F4, RNG} <:
+               AbstractSSAJumpAggregator
+    next_jump::Int
+    prev_jump::Int
+    next_jump_time::T
+    end_time::T
+    cur_rates::Vector{T}
+    sum_rate::T
+    ma_jumps::S
+    rate_bnds::F3
+    wds::F4
+    rates::F1
+    affects!::F2
+    save_positions::Tuple{Bool, Bool}
+    rng::RNG
+end
+
+function ExtrandeJumpAggregation(nj::Int, njt::T, et::T, crs::Vector{T}, sr::T, maj::S,
+                                 rs::F1, affs!::F2, sps::Tuple{Bool, Bool}, rng::RNG;
+                                 rate_bounds::F3, windows::F4,
+                                 kwargs...) where {T, S, F1, F2, F3, F4, RNG}
+    ExtrandeJumpAggregation{T, S, F1, F2, F3, F4, RNG}(nj, nj, njt, et, crs, sr, maj,
+                                                       rate_bounds, windows, rs, affs!, sps,
+                                                       rng)
+end
+
+############################# Required Functions ##############################
+function aggregate(aggregator::Extrande, u, p, t, end_time, constant_jumps,
+                   ma_jumps, save_positions, rng; variable_jumps = (), kwargs...)
+    rates, affects! = get_jump_info_fwrappers(u, p, t,
+                                              (constant_jumps..., variable_jumps...,
+                                               NullAffectJump))
+    rbnds, wnds = get_va_jump_bound_info_fwrapper(u, p, t,
+                                                  (constant_jumps..., variable_jumps...,
+                                                   NullAffectJump))
+    build_jump_aggregation(ExtrandeJumpAggregation, u, p, t, end_time, ma_jumps,
+                           rates, affects!, save_positions, rng; u = u, rate_bounds = rbnds,
+                           windows = wnds, kwargs...)
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::ExtrandeJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    generate_jumps!(p, integrator, u, params, t)
+end
+
+# execute one jump, changing the system state
+@inline function execute_jumps!(p::ExtrandeJumpAggregation, integrator, u, params, t)
+    # execute jump
+    u = update_state!(p, integrator, u)
+    nothing
+end
+
+@fastmath function next_ma_jump(p::ExtrandeJumpAggregation, u, params, t)
+    ttnj = typemax(typeof(t))
+    nextrx = zero(Int)
+    majumps = p.ma_jumps
+    @inbounds for i in 1:get_num_majumps(majumps)
+        p.cur_rates[i] = evalrxrate(u, i, majumps)
+        dt = randexp(p.rng) / p.cur_rates[i]
+        if dt < ttnj
+            ttnj = dt
+            nextrx = i
+        end
+    end
+    nextrx, ttnj
```

Comment:
I'm a little confused, as the extrande method does not generate `ttnj` for individual reactions, but instead calculates the total propensity like the `Direct` method. What you are doing here seems more like Gillespie's first reaction method.

Reply:
Yes, the Gillespie's stuff shouldn't have stayed there.